src/libraries/System.IO.Compression/src/System/IO/Compression/ZipHelper.cs

```diff
@@ -207,26 +207,35 @@ internal static byte[] GetEncodedTruncatedBytesFromString(string? text, Encoding
         if (isUTF8 && encoding.GetMaxByteCount(text.Length) > maxBytes)
         {
             int totalCodePoints = 0;
+            int charIndex = 0;
             foreach (Rune rune in text.EnumerateRunes())
             {
                 if (totalCodePoints + rune.Utf8SequenceLength > maxBytes)
                 {
                     break;
                 }
                 totalCodePoints += rune.Utf8SequenceLength;
+                charIndex += rune.Utf16SequenceLength;
             }
 
-            bytes = encoding.GetBytes(text);
+            byte[] rentedArray = ArrayPool<byte>.Shared.Rent(encoding.GetMaxByteCount(text.Length));
```

Comment:
I'm missing it... why can't we just allocate the array initially to be of length totalCodePoints and decode into that?